homeassistant/components/esphome/alarm_control_panel.py

```diff
@@ -0,0 +1,160 @@
+    AlarmControlPanelState, str
+] = EsphomeEnumMapper(
+    {
+        AlarmControlPanelState.DISARMED: STATE_ALARM_DISARMED,
+        AlarmControlPanelState.ARMED_HOME: STATE_ALARM_ARMED_HOME,
+        AlarmControlPanelState.ARMED_AWAY: STATE_ALARM_ARMED_AWAY,
+        AlarmControlPanelState.ARMED_NIGHT: STATE_ALARM_ARMED_NIGHT,
+        AlarmControlPanelState.ARMED_VACATION: STATE_ALARM_ARMED_VACATION,
+        AlarmControlPanelState.ARMED_CUSTOM_BYPASS: STATE_ALARM_ARMED_CUSTOM_BYPASS,
+        AlarmControlPanelState.PENDING: STATE_ALARM_PENDING,
+        AlarmControlPanelState.ARMING: STATE_ALARM_ARMING,
+        AlarmControlPanelState.DISARMING: STATE_ALARM_DISARMING,
+        AlarmControlPanelState.TRIGGERED: STATE_ALARM_TRIGGERED,
+    }
+)
+
+
+class EspHomeACPFeatures(APIIntEnum):
+    """ESPHome AlarmCintolPanel feature numbers."""
+
+    ARM_HOME = 1
+    ARM_AWAY = 2
+    ARM_NIGHT = 4
+    TRIGGER = 8
+    ARM_CUSTOM_BYPASS = 16
+    ARM_VACATION = 32
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up ESPHome switches based on a config entry."""
+    await platform_async_setup_entry(
+        hass,
+        entry,
+        async_add_entities,
+        component_key="alarm_control_panel",
+        info_type=AlarmControlPanelInfo,
+        entity_type=EsphomeAlarmControlPanel,
+        state_type=AlarmControlPanelEntityState,
+    )
+
+
+class EsphomeAlarmControlPanel(
+    EsphomeEntity[AlarmControlPanelInfo, AlarmControlPanelEntityState],
+    AlarmControlPanelEntity,
+):
+    """An Alarm Control Panel implementation for ESPHome."""
+
+    @property
+    def state(self) -> str | None:
+        """Return the state of the device."""
+        return _ESPHOME_ACP_STATE_TO_HASS_STATE.from_esphome(self._state.state)
+
+    @property
+    def supported_features(self) -> AlarmControlPanelEntityFeature:
+        """Return the list of supported features."""
+        feature = 0
+        if self._static_info.supported_features & EspHomeACPFeatures.ARM_HOME:
+            feature |= AlarmControlPanelEntityFeature.ARM_HOME
+        if self._static_info.supported_features & EspHomeACPFeatures.ARM_AWAY:
+            feature |= AlarmControlPanelEntityFeature.ARM_AWAY
+        if self._static_info.supported_features & EspHomeACPFeatures.ARM_NIGHT:
+            feature |= AlarmControlPanelEntityFeature.ARM_NIGHT
+        if self._static_info.supported_features & EspHomeACPFeatures.TRIGGER:
+            feature |= AlarmControlPanelEntityFeature.TRIGGER
+        if self._static_info.supported_features & EspHomeACPFeatures.ARM_CUSTOM_BYPASS:
+            feature |= AlarmControlPanelEntityFeature.ARM_CUSTOM_BYPASS
+        if self._static_info.supported_features & EspHomeACPFeatures.ARM_VACATION:
+            feature |= AlarmControlPanelEntityFeature.ARM_VACATION
+        return AlarmControlPanelEntityFeature(feature)
```

Comment:
Can you implement this as `self._attr_supported_features` in `__init__` (you'll need to make one call super()` so its not building `AlarmControlPanelEntityFeature` every time state is written?

The other entities need this as well but thats a discussion for another PR.

Reply:
I could not implement in `__init__` as the `_static_info` was not set yet, so I implemented a lazy load approach, but this does mean if the user changes the config in ESPHome the integration needs to be reloaded in order to get the updated features...

Reply:
```
    @property
    def _static_info(self) -> _InfoT:
        # Check if value is in info database. Use a single lookup.
        info = self._entry_data.info[self._component_key].get(self._key)
        if info is not None:
            return cast(_InfoT, info)
        # This entity is in the removal project and has been removed from .info
        # already, look in old_info
        return cast(_InfoT, self._entry_data.old_info[self._component_key][self._key])
```

That is a bit messy.

Ideally that would be set on the entity and available when its created. 

That's outside the scope of this PR though

Reply:
It looks like we actually have all the data when the entity is created.

```
        # First copy the now-old info into the backup object
        entry_data.old_info[component_key] = entry_data.info[component_key]
        # Then update the actual info
        entry_data.info[component_key] = new_infos

        # Add entities to Home Assistant
        async_add_entities(add_entities)
```

Reply:
I added a method in https://github.com/home-assistant/core/pull/94876 `_on_static_info_update` that could could override to know when the static info changes

Reply:
@grahambrown11 

Let's revert the lazy load approach for now so we can get this merged.
  
Sorry for the back and forth.  We can make it better after #94876

Reply:
Thanks for the feedback.
I'm not very familiar with Pyhon - I got an error when trying the `__init__` route so opted for the lazy load way... could I ask for assistance on what the __init__` should look like...
Should I wait for your PR to add the `_on_static_info_update` or put it on my todo list and create to a new PR later?

Reply:
Let's refactor in a future PR once #94876 merges since the non-lazy load version is consistent with the other platforms and it will be much easier to solve this in `__init__` + overriding the new super function once #94876 merges

Reply:
K, I've pushed the prev non-lazy load code.
No worries about the to and fro - we all change our minds 😆

Reply:
Both PRs are merged so you should be unblocked to do a follow up now

Reply:
I started working on a PR for sensor https://github.com/home-assistant/core/pull/94930. You'll probably want to wait to see the review feedback before starting on refactoring this